File: substrate/highlight_source_attach_quality_interrogation_compose.py
"""Highlight → floating DR + source attach quality interrogation (pure).

live_dispatched / merge_executed / remote_fetched / pdf_view_authorized
always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.highlight_deep_research_launch_compose import (
    HighlightDeepResearchLaunchCompose,
    HighlightDeepResearchLaunchComposeError,
    compose_highlight_deep_research_launch,
)
from substrate.source_attach_quality_interrogation_compose import (
    SourceAttachQualityInterrogationCompose,
    SourceAttachQualityInterrogationComposeError,
    compose_source_attach_quality_interrogation,
)


class HighlightSourceAttachQualityInterrogationComposeError(ValueError):
    """Fail-closed validation for highlight source attach interrogation pack."""


@dataclass(frozen=True)
class HighlightSourceAttachQualityInterrogationCompose:
    parent_asset_id: str
    session_id: str
    highlight_launch: HighlightDeepResearchLaunchCompose
    source_interrogation: SourceAttachQualityInterrogationCompose
    pack_ready: bool
    live_dispatched: bool
    merge_executed: bool
    remote_fetched: bool
    pdf_view_authorized: bool
    record_persisted: bool
    prompts_injected: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "session_id": self.session_id,
            "highlight_launch": self.highlight_launch.to_dict(),
            "source_interrogation": self.source_interrogation.to_dict(),
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "merge_executed": False,
            "remote_fetched": False,
            "pdf_view_authorized": False,
            "record_persisted": False,
            "prompts_injected": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": (
                "highlight_source_attach_quality_interrogation_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise HighlightSourceAttachQualityInterrogationComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_highlight_source_attach_quality_interrogation(
    *,
    parent_asset_id: object,
    highlight: object,
    gated: object,
    would_exceed: object,
    operator_ack: object,
    session_id: object,
    requested_families: object,
    sources: object,
    quality_overall: object,
    questions: object,
    chase_mode: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    prompt: object | None = None,
    preferred_view_mode: object | None = None,
    operator_override: object | None = None,
    selected_model_id: object | None = None,
    source_families: object | None = None,
    citations: object | None = None,
    derive_citations_from_sources: object | None = None,
    quality_floor: object | None = None,
    prior_records: object | None = None,
    user_prompt: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
    require_both: object | None = None,
) -> HighlightSourceAttachQualityInterrogationCompose:
    """Highlight DR launch + source attach interrogation. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise HighlightSourceAttachQualityInterrogationComposeError(
            "operator_ack must be an explicit boolean"
        )
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    session = _require_nonempty(session_id, field="session_id")
    hl = _require_nonempty(highlight, field="highlight")

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise HighlightSourceAttachQualityInterrogationComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false — highlight launch + interrogation pure intent",
        "merge_executed=false — parent asset not mutated",
        "remote_fetched=false — no live arxiv/substack scrape",
        "pdf_view_authorized=false — HTML-native only",
    ]

    selected = selected_model_id
    if selected is None or (isinstance(selected, str) and not selected.strip()):
        if not isinstance(models, list) or len(models) == 0:
            raise HighlightSourceAttachQualityInterrogationComposeError(
                "selected_model_id or models[0] required"
            )
        first = models[0]
        if not isinstance(first, dict):
            raise HighlightSourceAttachQualityInterrogationComposeError(
                "models[0] must be an object"
            )
        selected = _require_nonempty(
            first.get("model_id"), field="models[0].model_id"
        )
    else:
        selected = _require_nonempty(selected, field="selected_model_id")

    if user_prompt is not None and str(user_prompt).strip():
        uprompt = _require_nonempty(user_prompt, field="user_prompt")
    else:
        uprompt = hl

    launch_families = source_families
    if launch_families is None:
        launch_families = requested_families

    try:
        highlight_launch = compose_highlight_deep_research_launch(
            parent_asset_id=parent,
            highlight=hl,
            gated=gated,
            would_exceed=would_exceed,
            operator_ack=operator_ack,
            prompt=prompt if prompt is not None else uprompt,
            preferred_view_mode=preferred_view_mode,
            operator_override=operator_override,
            selected_model_id=selected,
            source_families=launch_families,
        )
    except HighlightDeepResearchLaunchComposeError as e:
        raise HighlightSourceAttachQualityInterrogationComposeError(str(e)) from e
    notes.extend(f"[highlight_launch] {n}" for n in highlight_launch.notes)

    try:
        source_interrogation = compose_source_attach_quality_interrogation(
            session_id=session,
            parent_asset_id=parent,
            requested_families=requested_families,
            sources=sources,
            quality_overall=quality_overall,
            would_exceed=would_exceed,
            questions=questions,
            chase_mode=chase_mode,
            user_prompt=uprompt,
            selected_model_id=selected,
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            operator_ack=operator_ack,
            citations=citations,
            derive_citations_from_sources=derive_citations_from_sources,
            quality_floor=quality_floor,
            operator_override=operator_override,
            prior_records=prior_records,
            projected_cost_usd_high=projected_cost_usd_high,
            projected_cost_usd_low=projected_cost_usd_low,
            bench_bests=bench_bests,
            focus_task=(
                "deep_research" if focus_task is None else focus_task
            ),
            nd_shadow=nd_shadow,
            require_both=True,
        )
    except SourceAttachQualityInterrogationComposeError as e:
        raise HighlightSourceAttachQualityInterrogationComposeError(str(e)) from e
    notes.extend(
        f"[source_interrogation] {n}" for n in source_interrogation.notes
    )

    if require:
        pack_ready = (
            highlight_launch.launch_ready is True
            and source_interrogation.pack_ready is True
            and operator_ack is True
        )
    else:
        pack_ready = operator_ack is True and (
            highlight_launch.launch_ready is True
            or source_interrogation.pack_ready is True
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — highlight launch + source interrogation ready; "
            "still pure"
        )
    else:
        notes.append(
            "pack_ready=false — launch, source interrogation, or operator_ack "
            "gate open"
        )

    if (
        highlight_launch.live_dispatched is not False
        or highlight_launch.merge_executed is not False
        or source_interrogation.remote_fetched is not False
        or source_interrogation.live_dispatched is not False
        or source_interrogation.record_persisted is not False
        or source_interrogation.prompts_injected is not False
    ):
        raise HighlightSourceAttachQualityInterrogationComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "merge_executed=false",
            "remote_fetched=false",
            "pdf_view_authorized=false",
            "record_persisted=false",
            "prompts_injected=false",
            "store_mutated=false",
        )
    )

    return HighlightSourceAttachQualityInterrogationCompose(
        parent_asset_id=parent,
        session_id=session,
        highlight_launch=highlight_launch,
        source_interrogation=source_interrogation,
        pack_ready=pack_ready,
        live_dispatched=False,
        merge_executed=False,
        remote_fetched=False,
        pdf_view_authorized=False,
        record_persisted=False,
        prompts_injected=False,
        store_mutated=False,
        notes=tuple(notes),
        authority=(
            "highlight_source_attach_quality_interrogation_compose_advisory"
        ),
    )


def format_highlight_source_attach_quality_interrogation_summary(
    c: HighlightSourceAttachQualityInterrogationCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"launch_ready={c.highlight_launch.launch_ready} · "
        f"source_ready={c.source_interrogation.pack_ready} · "
        f"live_dispatched=false · merge_executed=false · remote_fetched=false"
    )


__all__ = [
    "HighlightSourceAttachQualityInterrogationCompose",
    "HighlightSourceAttachQualityInterrogationComposeError",
    "compose_highlight_source_attach_quality_interrogation",
    "format_highlight_source_attach_quality_interrogation_summary",
]

File: apps/reading/src/api/highlightSourceAttachQualityInterrogationCompose.ts
/**
 * Highlight → floating DR launch + source attach quality interrogation (pure).
 *
 * Operator vision: from a reading/research highlight, spin up floating deep
 * research and attach arxiv/substack HTML sources under quality/budget gates
 * with workstation interrogation chase — same craft as research workstation
 * because reading and research share the HTML asset path.
 *
 * live_dispatched / merge_executed / remote_fetched / pdf_view_authorized
 * always false.
 */

import {
  composeHighlightDeepResearchLaunch,
  type HighlightDeepResearchLaunchCompose,
  type LaunchPreferredView,
  type SourceFamilyHint,
} from "./highlightDeepResearchLaunchCompose";
import {
  composeSourceAttachQualityInterrogation,
  type SourceAttachQualityInterrogationCompose,
} from "./sourceAttachQualityInterrogationCompose";
import type { HtmlNativeSourceRef } from "./htmlNativeSourceAttachCompose";
import type { CitationRecord } from "./deepResearchSourceCitationPack";
import type { PublicationFamily } from "./sourcePublicationRegistry";
import type {
  ChaseMode,
  ChaseQuestion,
} from "./researchInterrogationSubagentChaseCompose";
import type { SessionRecordItem } from "./workstationSessionInsightRecordCompose";
import type { ModelOption } from "./modelDecisionPromptCompose";
import type {
  BenchTaskBest,
  NotDiamondShadowRec,
} from "./settingsModelDriverTabCompose";

export type { LaunchPreferredView, SourceFamilyHint };

export interface HighlightSourceAttachQualityInterrogationInput {
  parent_asset_id: string;
  highlight: string;
  gated: boolean;
  prompt?: string;
  preferred_view_mode?: LaunchPreferredView;
  would_exceed: boolean | null;
  operator_override?: boolean;
  selected_model_id?: string | null;
  source_families?: SourceFamilyHint[] | null;
  operator_ack: boolean;
  /** Session for source attach / interrogation. */
  session_id: string;
  requested_families: PublicationFamily[];
  sources: HtmlNativeSourceRef[];
  citations?: CitationRecord[] | null;
  derive_citations_from_sources?: boolean;
  quality_overall: number | null;
  quality_floor?: number;
  questions: ChaseQuestion[];
  chase_mode: ChaseMode;
  prior_records?: SessionRecordItem[] | null;
  user_prompt?: string | null;
  models: ModelOption[];
  daily_cap_usd: number | null;
  spent_usd: number | null;
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  bench_bests?: BenchTaskBest[] | null;
  focus_task?: string | null;
  nd_shadow?: NotDiamondShadowRec | null;
  require_both?: boolean;
}

export interface HighlightSourceAttachQualityInterrogationCompose {
  parent_asset_id: string;
  session_id: string;
  highlight_launch: HighlightDeepResearchLaunchCompose;
  source_interrogation: SourceAttachQualityInterrogationCompose;
  pack_ready: boolean;
  live_dispatched: false;
  merge_executed: false;
  remote_fetched: false;
  pdf_view_authorized: false;
  record_persisted: false;
  prompts_injected: false;
  store_mutated: false;
  notes: string[];
  authority: "highlight_source_attach_quality_interrogation_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose highlight DR launch + source attach quality interrogation.
 * Never dispatches; never merges; never scrapes.
 */
export function composeHighlightSourceAttachQualityInterrogation(
  input: HighlightSourceAttachQualityInterrogationInput,
): HighlightSourceAttachQualityInterrogationCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  const session_id = requireNonEmpty(input.session_id, "session_id");

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false — highlight launch + interrogation pure intent",
    "merge_executed=false — parent asset not mutated",
    "remote_fetched=false — no live arxiv/substack scrape",
    "pdf_view_authorized=false — HTML-native only",
  ];

  // Prefer explicit selected_model_id for both paths; default first model if set.
  let selected_model_id =
    input.selected_model_id != null && input.selected_model_id !== undefined
      ? requireNonEmpty(input.selected_model_id, "selected_model_id")
      : null;
  if (selected_model_id == null && input.models.length > 0) {
    selected_model_id = requireNonEmpty(
      input.models[0].model_id,
      "models[0].model_id",
    );
  }
  if (selected_model_id == null) {
    throw new Error("selected_model_id or models[0] required");
  }

  const user_prompt =
    input.user_prompt != null && input.user_prompt !== undefined
      ? requireNonEmpty(input.user_prompt, "user_prompt")
      : requireNonEmpty(input.highlight, "highlight");

  // Merge source_families hints: prefer requested_families for attach path.
  const launch_families: SourceFamilyHint[] =
    input.source_families != null && input.source_families.length > 0
      ? input.source_families
      : (input.requested_families as SourceFamilyHint[]);

  const highlight_launch = composeHighlightDeepResearchLaunch({
    parent_asset_id,
    highlight: input.highlight,
    gated: input.gated,
    prompt: input.prompt ?? user_prompt,
    preferred_view_mode: input.preferred_view_mode,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    selected_model_id,
    source_families: launch_families,
    operator_ack: input.operator_ack,
  });
  notes.push(...highlight_launch.notes.map((n) => `[highlight_launch] ${n}`));

  const source_interrogation = composeSourceAttachQualityInterrogation({
    session_id,
    parent_asset_id,
    requested_families: input.requested_families,
    sources: input.sources,
    citations: input.citations,
    derive_citations_from_sources: input.derive_citations_from_sources,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    questions: input.questions,
    chase_mode: input.chase_mode,
    prior_records: input.prior_records,
    user_prompt,
    selected_model_id,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
    bench_bests: input.bench_bests,
    focus_task: input.focus_task ?? "deep_research",
    nd_shadow: input.nd_shadow,
    operator_ack: input.operator_ack,
    require_both: true,
  });
  notes.push(
    ...source_interrogation.notes.map((n) => `[source_interrogation] ${n}`),
  );

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      highlight_launch.launch_ready === true &&
      source_interrogation.pack_ready === true &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      (highlight_launch.launch_ready === true ||
        source_interrogation.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — highlight launch + source interrogation ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — launch, source interrogation, or operator_ack gate open",
    );
  }

  if (
    highlight_launch.live_dispatched !== false ||
    highlight_launch.merge_executed !== false ||
    source_interrogation.remote_fetched !== false ||
    source_interrogation.live_dispatched !== false ||
    source_interrogation.record_persisted !== false ||
    source_interrogation.prompts_injected !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("live_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("remote_fetched=false");
  notes.push("pdf_view_authorized=false");
  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");
  notes.push("store_mutated=false");

  return {
    parent_asset_id,
    session_id,
    highlight_launch,
    source_interrogation,
    pack_ready,
    live_dispatched: false,
    merge_executed: false,
    remote_fetched: false,
    pdf_view_authorized: false,
    record_persisted: false,
    prompts_injected: false,
    store_mutated: false,
    notes,
    authority: "highlight_source_attach_quality_interrogation_compose_advisory",
  };
}

export function formatHighlightSourceAttachQualityInterrogationSummary(
  c: HighlightSourceAttachQualityInterrogationCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `launch_ready=${c.highlight_launch.launch_ready} · ` +
    `source_ready=${c.source_interrogation.pack_ready} · ` +
    `live_dispatched=false · merge_executed=false · remote_fetched=false`
  );
}
